bitcoin/src/bip32.rs

```diff
@@ -373,103 +620,209 @@ impl<'a> Iterator for DerivationPathIterator<'a> {
+            NormalDerivationPath::try_from(self.0.as_slice())
+        }
+        /// Check if the derivation path is empty
+        pub fn is_empty(&self) -> bool { self.0.is_empty() }
+        /// Returns length of the derivation path
+        pub fn len(&self) -> usize { self.0.len() }
+
+        /// Returns whether derivation path represents master key (i.e. it's length
+        /// is empty). True for `m` path.
+        pub fn is_master(&self) -> bool { self.0.is_empty() }
+
+        /// Create a new [DerivationPath] that is a child of this one.
+        pub fn child(&self, cn: ChildNumber) -> Self {
+            self.into_iter().chain(core::iter::once(&cn)).collect()
+        }
 
-    /// Returns derivation path for a master key (i.e. empty derivation path)
-    pub fn master() -> DerivationPath { DerivationPath(vec![]) }
+        /// Convert into a [DerivationPath] that is a child of this one.
+        pub fn into_child(self, cn: ChildNumber) -> Self {
+            let mut path = self.0;
+            path.push(cn);
+            DerivationPath::new(path)
+        }
 
-    /// Returns whether derivation path represents master key (i.e. it's length
-    /// is empty). True for `m` path.
-    pub fn is_master(&self) -> bool { self.0.is_empty() }
+        /// Get an [Iterator] over the children of this [DerivationPath]
+        /// starting with the given [ChildNumber].
+        pub fn children_from(&self, cn: ChildNumber) -> DerivationPathIterator {
+            DerivationPathIterator::start_from(self, cn)
+        }
 
-    /// Create a new [DerivationPath] that is a child of this one.
-    pub fn child(&self, cn: ChildNumber) -> DerivationPath {
-        let mut path = self.0.clone();
-        path.push(cn);
-        DerivationPath(path)
+        /// Get an [Iterator] over the unhardened children of this [DerivationPath].
+        pub fn normal_children(&self) -> DerivationPathIterator {
+            DerivationPathIterator::start_from(self, ChildNumber::ZERO_NORMAL)
+        }
+
+        /// Get an [Iterator] over the hardened children of this [DerivationPath].
+        pub fn hardened_children(&self) -> DerivationPathIterator {
+            DerivationPathIterator::start_from(self, ChildNumber::ZERO_HARDENED)
+        }
+
+        /// Concatenate `self` with `path` and return the resulting new path.
+        ///
+        /// ```
+        /// use bitcoin::bip32::{DerivationPath, ChildNumber};
+        /// use std::str::FromStr;
+        ///
+        /// let base = DerivationPath::from_str("m/42").unwrap();
+        ///
+        /// let deriv_1 = base.extend(DerivationPath::from_str("0/1").unwrap());
+        /// let deriv_2 = base.extend(&[
+        ///     ChildNumber::ZERO_NORMAL,
+        ///     ChildNumber::ONE_NORMAL
+        /// ]);
+        ///
+        /// assert_eq!(deriv_1, deriv_2);
+        /// ```
+        pub fn extend<T: AsRef<[ChildNumber]>>(&self, path: T) -> Self {
```

Comment:
This should either mutate `self` (but then it should use `Extend` trait) or have `#[must_use]` attribute.

Reply:
Yep, but the docs of the function specifies to return a new item, did like this to preserve the original purpose of the function.